packages/video_player/video_player_avfoundation/example/lib/mini_controller.dart

```diff
@@ -243,6 +273,9 @@ class MiniController extends ValueNotifier<VideoPlayerValue> {
         case VideoEventType.bufferingEnd:
           value = value.copyWith(isBuffering: false);
           break;
+        case VideoEventType.isPlayingStateUpdate:
+          value = value.copyWith(isPlaying: event.isPlaying);
```

Comment:
noob question: is `copyWith` functionality enabled by `immutable`?

Reply:
[immutable top-level constant](https://api.flutter.dev/flutter/meta/immutable-constant.html)

> A class is immutable if all of the instance fields of the class, whether defined directly or inherited, are final.

While the `@immutable` annotation is commonly used to indicate that a class is immutable in Dart, it does not automatically provide a `copyWith` method for the class. However, defining a `copyWith` method is a common pattern for creating new instances of immutable objects with modified values. Note that some classes like `Widget` or `StatelessWidget` in Flutter may have the `@immutable` annotation, but they do not necessarily provide a `copyWith` method since they are more like blueprints or configurations rather than data models.

Since `VideoPlayerValue` already using `final` fields it's easy to add `@immutable`.

Since `VideoPlayerController extends ValueNotifier<VideoPlayerValue>` we have to create a new a `value` and since `ValueNotifier` notifies it's listeners when it gets a new `value` it should override `equals` method.

[ValueNotifier<T> class](https://api.flutter.dev/flutter/foundation/ValueNotifier-class.html)

> A [ChangeNotifier](https://api.flutter.dev/flutter/foundation/ChangeNotifier-class.html) that holds a single value.
> 
> When [value](https://api.flutter.dev/flutter/foundation/ValueNotifier/value.html) is replaced with something that is not equal to the old value as evaluated by the equality operator ==, this class notifies its listeners.